streaming/base/format/pq/reader.py

```diff
@@ -0,0 +1,188 @@
+        mds_filename = os.path.join(self.dirname, self.split, self.raw_data.basename)
+        if os.path.exists(mds_filename):
+            if os.path.exists(pq_filename):
+                if safe_keep_zip:
+                    # Present: keep both (because of safe_keep_zip).
+                    usage = os.stat(mds_filename).st_size + os.stat(pq_filename).st_size
+                else:
+                    # Present: keep MDS, drop PQ (because of safe_keep_zip).
+                    os.remove(pq_filename)
+                    usage = os.stat(mds_filename).st_size
+            else:
+                if safe_keep_zip:
+                    # Normalize to missing, because safe_keep_zip requires that we keep the PQ.
+                    os.remove(mds_filename)
+                    usage = 0
+                else:
+                    # Present: have MDS, don't have or want PQ.
+                    usage = os.stat(mds_filename).st_size
+        else:
+            if os.path.exists(pq_filename):
+                # Present: PQ hasn't been converted to MDS yet and we don't have time to here.
+                usage = os.stat(pq_filename).st_size
+            else:
+                # Missing: both PQ and MDS are not there.
+                usage = 0
+        return usage
+
+    def get_column(self, val: Any) -> str:
+        """Get the MDS column encoding of one field.
+
+        Args:
+            val (Any): The field.
+
+        Returns:
+            str: Its corresponding MDS encoding.
+        """
+        if isinstance(val, int):
+            return 'int'
+        elif isinstance(val, str):
+            return 'str'
+        else:
+            raise ValueError('Unsupported column type: {type(val)}.')
+
+    def get_columns(self, sample: Dict[str, Any]) -> Dict[str, str]:
+        """Get the MDS columns given one sample.
+
+        Args:
+            sample (Dict[str, Any]): Mapping of column name to value.
+
+        Returns:
+            Dict[str, str]: Mapping of column name to MDS encoding.
+        """
+        col_names = sorted(sample)
+        col_encs = []
+        for name in col_names:
+            val = sample[name]
+            enc = self.get_column(val)
+            col_encs.append(enc)
+        return dict(zip(col_names, col_encs))
+
+    def prepare(self, safe_keep_zip: bool) -> int:
+        """Prepare a Parquet shard for fast random access by converting to MDS.
+
+        Args:
+            safe_keep_zip (bool): Whether to keep Parquet shards, or drop post-conversion.
+
+        Returns:
+            int: Change in cache usage in bytes due to PQ -> MDS conversion.
+        """
+        pq_filename = os.path.join(self.dirname, self.split, self.raw_parquet.basename)
+        table = pq.read_table(pq_filename)
+        samples = table.to_pylist()
+        columns = self.get_columns(samples[0])
+        with TemporaryDirectory() as temp_dir:
+            with MDSWriter(columns=columns, out=temp_dir, size_limit=None) as out:
```

Comment:
This can be a compound with statement for clarity:
```suggestion
        with TemporaryDirectory() as temp_dir, MDSWriter(columns=columns, out=temp_dir, size_limit=None) as out:
```

Reply:
They close at different times